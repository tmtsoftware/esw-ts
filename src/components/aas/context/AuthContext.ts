import { createContext } from 'react'
import type { Auth } from '../../../clients/aas'

export interface AuthContextType {
  auth: Auth | null
  login: () => void
  logout: () => void
}

const AuthContextDefaultState = {
  auth: null,
  login: () => undefined,
  logout: () => undefined
}
/**
 * Default state for AuthContextProvider
 * @type {{auth: undefined, login: (function(): boolean), logout: (function(): boolean)}}
 */
const AuthContext = createContext<AuthContextType>(AuthContextDefaultState)
const { Provider, Consumer } = AuthContext

// todo: AuthContext is exported to support scala.js, see if only exporting Consumer works
export { AuthContext, Provider, Consumer, AuthContextDefaultState }
