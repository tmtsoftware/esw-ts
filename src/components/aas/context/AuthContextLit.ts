/*
 * Copyright (C) 2023 Thirty Meter Telescope International Observatory
 * SPDX-License-Identifier: Apache-2.0
 */

import { type Auth, type AuthContextConfig, AuthStore } from '../../../clients/aas'

export class AuthContextLit {
  auth: Auth | null = null
  realm: string = 'TMT'
  clientId: string = 'tmt-frontend-app'

  /**
   * Instantiate keycloak and sets AuthStore instance in state. This state can be provided
   * as a context
   */
  private async instantiateAAS(url: string, redirect: boolean) {
    const keycloakConfig: AuthContextConfig = { realm: this.realm, clientId: this.clientId }

    const { keycloak, authenticatedPromise } = AuthStore.authenticate(keycloakConfig, url, redirect)
    try {
      await authenticatedPromise
      const _auth = AuthStore.from(keycloak)
      this.auth = _auth
    } catch (e) {
      console.error(e)
      this.auth = null
    }
  }

  /**
   * Resolves AAS server and instantiate keycloak in check-sso mode
   */
  async loginWithoutRedirect() {
    const url = await AuthStore.getAASUrl()
    await this.instantiateAAS(url, false)
  }

  /**
   * Resolves AAS server and instantiate keycloak in login-required mode
   */
  async login() {
    const url = await AuthStore.getAASUrl()
    console.log(url)
    await this.instantiateAAS(url, true)
  }

  async logout() {
    if (this.auth && this.auth.logout) {
      const logoutPromise = this.auth.logout()
      logoutPromise.then(() => {
        this.auth = null
      })
      await this.auth.logout()
    }
  }
}
