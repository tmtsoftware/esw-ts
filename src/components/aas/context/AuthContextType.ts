import type { Auth } from '../../../clients/aas'

export interface AuthContextType {
  auth: Auth | null
  login: () => void
  logout: () => void
}

export const AuthContextDefaultState = {
  auth: null,
  login: () => undefined,
  logout: () => undefined
}
