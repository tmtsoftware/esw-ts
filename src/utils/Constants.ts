export const APP_NAME = 'App-Name'
export const USERNAME = 'Username'
export const UNKNOWN_USERNAME = 'unknown'
