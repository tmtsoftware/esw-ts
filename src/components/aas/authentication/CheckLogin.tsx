import React, { useContext } from 'react'
import { AuthContext, AuthContextType } from '../context/AuthContext'

export interface CheckLoginProps {
  children: React.ReactNode
  error: React.ReactNode
}

/**
 * React component which renders children if authenticated and error if unauthenticated
 * @param children (optional prop) can be react components or html element which will be rendered
 * if user is authenticated
 * @param error (optional prop) can be react components or html element which will be rendered
 * if user is not authenticated
 */

const CheckLogin = ({ children, error }: CheckLoginProps) => {
  const { auth } = useContext<AuthContextType>(AuthContext)
  const node = auth && auth.isAuthenticated && auth.isAuthenticated() ? children : error
  return <div>{node}</div>
}

export default CheckLogin
