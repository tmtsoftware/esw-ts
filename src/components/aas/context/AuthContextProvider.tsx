import React, { useEffect, useState } from 'react'
import { Auth, AuthStore } from '../../../clients/aas'
import { Provider } from './AuthContext'

export interface AuthContextConfig {
  realm: string
  clientId: string
  username?: string
  password?: string
}

export interface AuthContextProps {
  config: AuthContextConfig
  children: React.ReactNode
}

// TODO Add unit tests
/**
 * React component which is wrapper over provider of react context api.
 * Responsible for instantiating keycloak and provide context value to consumers
 * props -
 * config json specific to UI application e.g. realm and clientId
 * children - react component or html element which can have consumer to access
 * context provided
 */
const AuthContextProvider = (props: AuthContextProps) => {
  const [auth, setAuth] = useState<Auth | null>(null)

  /**
   * Instantiate keycloak and sets AuthStore instance in state. This state can be provided
   * as a context
   */
  const instantiateAAS = async (url: string, redirect: boolean) => {
    const { keycloak, authenticatedPromise } = AuthStore.authenticate(props.config, url, redirect)
    try {
      await authenticatedPromise
      const _auth = AuthStore.from(keycloak)
      setAuth(_auth)
    } catch (e) {
      console.error(e)
      setAuth(null)
    }
  }

  /**
   * Resolves AAS server and instantiate keycloak in check-sso mode
   */
  const loginWithoutRedirect = async () => {
    const url = await AuthStore.getAASUrl()
    await instantiateAAS(url, false)
  }

  useEffect(() => {
    loginWithoutRedirect()
  }, [])

  /**
   * Resolves AAS server and instantiate keycloak in login-required mode
   */
  const login = async () => {
    const url = await AuthStore.getAASUrl()
    console.log(url)
    await instantiateAAS(url, true)
  }

  const logout = async () => {
    if (auth && auth.logout) {
      const logoutPromise = auth.logout()
      logoutPromise.then(() => {
        setAuth(null)
      })
      await auth.logout()
    }
  }

  return <Provider value={{ auth: auth, login: login, logout: logout }}>{props.children}</Provider>
}

export default AuthContextProvider
