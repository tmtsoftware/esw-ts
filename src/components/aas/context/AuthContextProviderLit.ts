/*
 * Copyright (C) 2023 Thirty Meter Telescope International Observatory
 * SPDX-License-Identifier: Apache-2.0
 */

import {createContext, provide} from '@lit/context'
import {LitElement, html} from 'lit'
import {customElement, property} from 'lit/decorators.js'
import {AuthContextLit} from "./AuthContextLit"

export const litAuthContext = createContext<AuthContextLit>('authContext')

@customElement('auth-context-provider')
export class AuthContextProviderLit extends LitElement {
  @property({type: String})
  realm: string = 'TMT'

  @property({attribute: 'client-id', type: String})
  clientId: string = 'tmt-frontend-app'

  @provide({context: litAuthContext})
  @property({attribute: false})
  authContext = new AuthContextLit()

  render() {
    return html`
      <div class="container">
        <slot></slot>
      </div>`
  }
}

declare global {
  interface HTMLElementTagNameMap {
    'auth-context-provider': AuthContextProviderLit
  }
}
