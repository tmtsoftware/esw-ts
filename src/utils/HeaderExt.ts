import { UNKNOWN_USERNAME, USERNAME } from './Constants'

export class HeaderExt extends Headers {
  constructor(values?: HeadersInit) {
    super(values)
  }

  //TODO add tests for the helper methods
  withHeader(headerName: string, value: string) {
    super.set(headerName, value)
    return new HeaderExt(this)
  }

  withContentType(value: string) {
    return this.withHeader('Content-Type', value)
  }

  withUsername(value?: string) {
    return this.withHeader(USERNAME, value ? value : UNKNOWN_USERNAME)
  }

  withAuthorization(token?: string) {
    if (!token) return this
    return this.withHeader('Authorization', `Bearer ${token}`)
  }
}
