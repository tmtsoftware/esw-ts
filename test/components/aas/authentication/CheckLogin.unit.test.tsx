import { render } from '@testing-library/react'
import Keycloak from 'keycloak-js'
import * as React from 'react'
import { mocked } from 'ts-jest/utils'
import CheckLogin from '../../../../src/components/aas/authentication/CheckLogin'
import { Provider } from '../../../../src/components/aas/context/AuthContext'
import { mockAuth, mockedKeyCloakInstance } from '../../../helpers/MockHelpers'

jest.mock('keycloak-js')

describe('<CheckLogin />', () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })

  const mockFn = mocked(Keycloak, true)
  const keycloakInstance = mockedKeyCloakInstance()
  mockFn.mockReturnValue(keycloakInstance)

  const props = {
    children: <div id="auth">Authentication successful</div>,
    error: <div id="error">Authentication unsuccessful</div>
  }
  const mockLogin = jest.fn()
  const mockLogout = jest.fn()

  test('should render children elements if authentication is true | ESW-330', async () => {
    const auth = mockAuth(true)

    const { queryByText } = render(
      <Provider value={{ auth: auth, login: mockLogin, logout: mockLogout }}>
        <CheckLogin {...props} />
      </Provider>
    )

    expect(queryByText('Authentication successful')).toBeTruthy()
    expect(queryByText('Authentication unsuccessful')).toBeFalsy()
  })

  test('should not render children elements if authentication is false', () => {
    const auth = mockAuth(false)

    const { queryByText } = render(
      <Provider value={{ auth: auth, login: mockLogin, logout: mockLogout }}>
        <CheckLogin {...props} />
      </Provider>
    )

    expect(queryByText('Authentication successful')).toBeFalsy()
    expect(queryByText('Authentication unsuccessful')).toBeTruthy()
  })

  test('should render CheckLogin if authentication is true', () => {
    const auth = mockAuth(true)

    const { container } = render(
      <Provider value={{ auth: auth, login: mockLogin, logout: mockLogout }}>
        <CheckLogin {...props} />
      </Provider>
    )

    expect(container).toMatchSnapshot()
  })

  test('should not render CheckLogin if authentication is false', () => {
    const auth = mockAuth(false)

    const { container } = render(
      <Provider value={{ auth: auth, login: mockLogin, logout: mockLogout }}>
        <CheckLogin {...props} />
      </Provider>
    )

    expect(container).toMatchSnapshot()
  })

  test('should render nothing if CheckLogin if authentication is false and error component is not provided', () => {
    const auth = mockAuth(false)

    const props = {
      children: <div className="auth">Authentication successful</div>,
      error: null
    }

    const { container } = render(
      <Provider value={{ auth: auth, login: mockLogin, logout: mockLogout }}>
        <CheckLogin {...props} />
      </Provider>
    )

    expect(container).toMatchSnapshot()
  })
})
