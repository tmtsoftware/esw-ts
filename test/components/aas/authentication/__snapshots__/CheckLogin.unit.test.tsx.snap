// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckLogin /> should not render CheckLogin if authentication is false 1`] = `
<div>
  <div>
    <div
      id="error"
    >
      Authentication unsuccessful
    </div>
  </div>
</div>
`;

exports[`<CheckLogin /> should render CheckLogin if authentication is true 1`] = `
<div>
  <div>
    <div
      id="auth"
    >
      Authentication successful
    </div>
  </div>
</div>
`;

exports[`<CheckLogin /> should render nothing if CheckLogin if authentication is false and error component is not provided 1`] = `
<div>
  <div />
</div>
`;
